### PR TITLE
Update to Jest 22

### PR DESCRIPTION
**Pull request by gaearon**
jsdom update made this one challenging.

<s>It uncovered some issues in tests, but I believe some of the remaining 9 tests I haven't fixed yet actually point out legit problems (where we emit a different number of warnings on the server and on the client).</s>

I made one change to the behavior. Previously we were emitting two different warnings (about bad prop casing and about client/server mismatch) when hydrating a badly cased SVG attribute like `<text textlength="10" />`. In https://github.com/facebook/react/pull/11956/commits/ed93325e007f76cbeefd11f5cf9c4c2b97bbd969, I changed it to only emit one warning (about the prop casing) since that's the actionable one. This fixed the tests (which assume the number of warnings is consistent between server and client).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -67,10 +67,7 @@
     "gzip-js": "~0.3.2",
     "gzip-size": "^3.0.0",
     "jasmine-check": "^1.0.0-rc.0",
-    "jest": "^21.3.0-beta.4",
-    "jest-config": "^21.3.0-beta.4",
-    "jest-jasmine2": "^21.3.0-beta.4",
-    "jest-runtime": "^21.3.0-beta.4",
```
Comment (gaearon) [inline package.json]: We weren't using these anymore

**Comment by jquense** [comment]
that SVG element support tho 👌

File: scripts/jest/setupEnvironment.js [hunk 2]
```diff
@@ -31,6 +31,20 @@ global.cancelIdleCallback = function(callbackID) {
...
+  // Same as above.
   DOMException.prototype.suppressReactErrorLogging = true;
+
+  // Also prevent JSDOM from logging intentionally thrown errors.
+  // TODO: it might make sense to do it the other way around.
+  // https://github.com/facebook/react/issues/11098#issuecomment-355032539
+  window.addEventListener('error', event => {
+    if (event.error != null && event.error.suppressReactErrorLogging) {
+      event.preventDefault();
+    }
```
Comment (jquense) [inline scripts/jest/setupEnvironment.js]: might be worth ensuring these are logged somewhere. I noticed on RDL that errors thrown in event handlers where being swallowed by jsdom under mysterious circumstances...I guess there is already an issue for it tho:  https://github.com/facebook/react/issues/8260

Comment (gaearon) [reply scripts/jest/setupEnvironment.js]: Yeah, thanks for highlighting that. They are logged now (by jsdom) but we should probably fail the build on them (?)

This issue (and related PR) is actually why I got curious about updating jsdom :P

**Comment by gaearon** [comment]
The remaining failing tests are in the DOM server integration suite and have to do with the fact that now we have a few cases where some warnings only fire on the client side (previously we didn't "see" them because of jsdom not faithfully emulating SVG or custom elements).

In some cases we can't really fix the discrepancy. For example "render HTML only and then use innerHTML" tests will never emit React warnings about unknown DOM element type because those only happen in the client-side code. So maybe we should explicitly check expected warnings somehow, or pass different numbers for different cases.

**Comment by gaearon** [comment]
Failing the build on uncaught errors in event handlers (https://github.com/facebook/react/issues/8260) is more challenging and I won't do it here. (We didn't do it before either.)

We know *how* to do it now though (now that we own the error handler), and I think that should be fixed as part of https://github.com/facebook/react/issues/11098#issuecomment-355032539.

File: packages/react-dom/src/__tests__/validateDOMNesting-test.js [hunk 3]
```diff
@@ -24,7 +24,13 @@ function expectWarnings(tags, warnings = []) {
   if (__DEV__) {
     console.error.calls.reset();
   }
-  const container = document.createElement(tags.splice(0, 1));
+
+  const containerTag = tags.shift();
+  const container =
+    containerTag === 'svg'
+      ? document.createElementNS('http://www.w3.org/2000/svg', containerTag)
```
Comment (bvaughn) [inline packages/react-dom/src/__tests__/validateDOMNesting-test.js]: Nifty

File: packages/react-dom/src/__tests__/ReactDOMServerIntegrationAttributes-test.js [hunk 4]
```diff
@@ -558,6 +590,16 @@ describe('ReactDOMServerIntegration', () => {
...
       itRenders('custom attributes for non-standard elements', async render => {
+        // This test suite generally assumes that we get exactly
+        // the same warnings (or none) for all scenarios including
+        // SSR + innerHTML, hydration, and client-side rendering.
+        // However this particular warning fires only when creating
+        // DOM nodes on the client side. We force it to fire early
+        // so that it gets deduplicated later, and doesn't fail the test.
+        expect(() => {
+          ReactDOM.render(<nonstandard />, document.createElement('div'));
+        }).toWarnDev('The tag <nonstandard> is unrecognized in this browser.');
```
Comment (bvaughn) [inline packages/react-dom/src/__tests__/ReactDOMServerIntegrationAttributes-test.js]: Clever solution. 👍

File: packages/react-dom/src/client/ReactDOMFiberComponent.js [hunk 5]
```diff
@@ -1035,6 +1048,17 @@ export function diffHydratedProperties(
...
+            const standardName = getPossibleStandardName(propKey);
+            if (standardName !== null && standardName !== propKey) {
+              // If an SVG prop is supplied with bad casing, it will
+              // be successfully parsed from HTML, but will produce a mismatch
+              // (and would be incorrectly rendered on the client).
+              // However, we already warn about bad casing elsewhere.
+              // So we'll skip the misleading extra mismatch warning in this case.
+              isMismatchDueToBadCasing = true;
+              // $FlowFixMe - Should be inferred as not undefined.
+              extraAttributeNames.delete(standardName);
```
Comment (bvaughn) [inline packages/react-dom/src/client/ReactDOMFiberComponent.js]: Nice!

**Comment by gaearon** [comment]
Yeah, forgot to update that part. Individual commits show how I fixed them.

File: package.json (hunk 1 above)
Comment (SimenB) [reply package.json]: `expect` is still in here in beta, is that used?

Comment (gaearon) [reply package.json]: Probably not..

**Pull request merged**